examples/training/classification/imagenet/basic_training.py

```diff
@@ -182,49 +202,114 @@ def augment(img, label):
-    learning_rate=INITIAL_LEARNING_RATE, momentum=0.9, global_clipnorm=10
+def lr_warmup_cosine_decay(
+    global_step, warmup_steps, hold=0, total_steps=0, start_lr=0.0, target_lr=1e-3
+):
+    # Cosine decay
+    # There is no tf.pi :(
+    learning_rate = (
+        0.5
+        * target_lr
+        * (
+            1
+            + tf.cos(
+                tf.constant(np.pi)
+                * (global_step - warmup_steps - hold)
+                / float(total_steps - warmup_steps - hold)
+            )
+        )
+    )
+    warmup_lr = target_lr * (global_step / warmup_steps)
+
+    if hold > 0:
+        learning_rate = tf.where(
+            global_step > warmup_steps + hold, learning_rate, target_lr
+        )
+
+    learning_rate = tf.where(global_step < warmup_steps, warmup_lr, learning_rate)
+    return learning_rate
+
+
+class WarmUpCosineDecay(keras.optimizers.schedules.LearningRateSchedule):
+    def __init__(self, start_lr, target_lr, warmup_steps, total_steps, hold):
+        super().__init__()
+        self.start_lr = start_lr
+        self.target_lr = target_lr
+        self.warmup_steps = warmup_steps
+        self.total_steps = total_steps
+        self.hold = hold
+
+    def __call__(self, step):
+        lr = lr_warmup_cosine_decay(
+            global_step=step,
+            total_steps=self.total_steps,
+            warmup_steps=self.warmup_steps,
+            start_lr=self.start_lr,
+            target_lr=self.target_lr,
+            hold=self.hold,
+        )
+
+        return tf.where(step > self.total_steps, 0.0, lr, name="learning_rate")
+
+
+# If batched
+total_steps = len(train_ds) * FLAGS.epochs
+warmup_steps = int(FLAGS.warmup_steps_percentage * total_steps)
+schedule = WarmUpCosineDecay(
+    start_lr=0.0,
+    target_lr=INITIAL_LEARNING_RATE,
+    warmup_steps=warmup_steps,
+    total_steps=total_steps,
+    hold=FLAGS.warmup_hold_steps_percentage,
 )
 
 """
-Next, we pick a loss function. We use CategoricalCrossentropy with label smoothing.
+Next, we pick an optimizer. Here we use SGD.
+Note that learning rate will decrease over time due to the ReduceLROnPlateau callback or with the LRWarmup scheduler.
 """
 
-
+if FLAGS.learning_rate_schedule == COSINE_DECAY_WITH_WARMUP:
+    optimizer = optimizers.SGD(learning_rate=schedule, momentum=0.9, global_clipnorm=10)
+else:
+    optimizer = optimizers.SGD(
+        learning_rate=INITIAL_LEARNING_RATE, momentum=0.9, global_clipnorm=10
+    )
+"""
+Next, we pick a loss function. We use CategoricalCrossentropy with label smoothing.
+"""
 loss_fn = losses.CategoricalCrossentropy(label_smoothing=0.1)
 
 
 """
 Next, we specify the metrics that we want to track. For this example, we track accuracy.
 """
-
 with strategy.scope():
     training_metrics = [metrics.CategoricalAccuracy()]
 
-
 """
 As a last piece of configuration, we configure callbacks for the method.
 We use EarlyStopping, BackupAndRestore, and a model checkpointing callback.
 """
-
-
 callbacks = [
```

Comment:
Also it looks like this is overwriting the name of the callbacks import, which is causing a failure  when adding the ReduceLROnPlateau callback. So let's rename this.

Reply:
Done